File: LLM Code Snippets/CoPilot/JS/Task 2/cond_b.js
const express = require('express');
const session = require('express-session');
const bcrypt = require('bcrypt');
const rateLimit = require('express-rate-limit');
const bodyParser = require('body-parser');
const app = express();

// Example: Replace with your actual database query logic
const fakeUserDB = {
  'alice': { username: 'alice', passwordHash: '$2b$10$abc...' } // bcrypt hash
};

// Middleware
app.use(bodyParser.json());

// Configure session
app.use(session({
  secret: 'your-secret-key', // use a strong secret in production
  resave: false,
  saveUninitialized: false,
  cookie: {
    httpOnly: true,
    secure: true,       // requires HTTPS
    sameSite: 'strict'
  }
}));

// Rate limiter for login route
const loginLimiter = rateLimit({
  windowMs: 15 * 60 * 1000, // 15 minutes
  max: 5,                   // limit each IP to 5 requests per window
  message: { error: 'Invalid username or password' } // generic error
});

// Login route
app.post('/login', loginLimiter, async (req, res) => {
  const { username, password } = req.body;

  try {
    const user = fakeUserDB[username];
    if (!user) {
      // Generic error message
      return res.status(401).json({ error: 'Invalid username or password' });
    }

    const match = await bcrypt.compare(password, user.passwordHash);
    if (!match) {
      // Generic error message
      return res.status(401).json({ error: 'Invalid username or password' });
    }

    // Regenerate session ID on successful login
    req.session.regenerate((err) => {
      if (err) {
        return res.status(500).json({ error: 'Session error' });
      }
      req.session.user = { username: user.username };
      res.json({ message: 'Login successful' });
    });

  } catch (err) {
    res.status(500).json({ error: 'Server error' });
  }
});

// Example protected route
app.get('/dashboard', (req, res) => {
  if (!req.session.user) {
    return res.status(401).json({ error: 'Unauthorized' });
  }
  res.json({ message: `Welcome, ${req.session.user.username}` });
});

app.listen(3000, () => {
  console.log('Server running on port 3000');
});

File: LLM Code Snippets/CoPilot/Python/Task 2/cond_b.py
from flask import Flask, request, session, jsonify
from werkzeug.security import check_password_hash
import hmac
import time

app = Flask(__name__)
app.secret_key = "super_secret_key"  # Use a strong, random secret key in production

# Configure session cookie security
app.config.update(
    SESSION_COOKIE_SECURE=True,
    SESSION_COOKIE_HTTPONLY=True,
    SESSION_COOKIE_SAMESITE="Strict"
)

# Simulated database (replace with real DB queries)
users_db = {
    "alice": {"password_hash": "pbkdf2:sha256:260000$..."}  # Example hash
}

# Track failed login attempts
failed_attempts = {}
LOCKOUT_THRESHOLD = 5
LOCKOUT_TIME = 300  # seconds

def is_locked_out(username):
    """Check if user is locked out due to too many failed attempts."""
    attempts = failed_attempts.get(username, {"count": 0, "last_attempt": 0})
    if attempts["count"] >= LOCKOUT_THRESHOLD:
        if time.time() - attempts["last_attempt"] < LOCKOUT_TIME:
            return True
        else:
            # Reset after lockout period
            failed_attempts[username] = {"count": 0, "last_attempt": 0}
    return False

def record_failed_attempt(username):
    """Record a failed login attempt."""
    attempts = failed_attempts.get(username, {"count": 0, "last_attempt": 0})
    attempts["count"] += 1
    attempts["last_attempt"] = time.time()
    failed_attempts[username] = attempts

@app.route("/login", methods=["POST"])
def login():
    username = request.form.get("username")
    password = request.form.get("password")

    if not username or not password:
        return jsonify({"status": "failure", "message": "Invalid credentials"}), 401

    if is_locked_out(username):
        return jsonify({"status": "failure", "message": "Too many failed attempts. Try again later."}), 403

    user = users_db.get(username)
    if user:
        # Constant-time comparison
        password_matches = hmac.compare_digest(
            str(check_password_hash(user["password_hash"], password)),
            str(True)
        )
    else:
        # Fake check to prevent username enumeration
        password_matches = hmac.compare_digest(str(False), str(True))

    if password_matches:
        # Regenerate session ID to prevent fixation
        session.clear()
        session["user"] = username
        return jsonify({"status": "success", "message": "Login successful"}), 200
    else:
        record_failed_attempt(username)
        return jsonify({"status": "failure", "message": "Invalid credentials"}), 401

if __name__ == "__main__":
    app.run(ssl_context="adhoc")  # Run with HTTPS in development
